src/graph.cpp

```diff
@@ -165,9 +165,66 @@ bool Graph::path(const std::list<std::string> &source, const std::string &target
     return false;
 }
 
+std::list<Edge *> Graph::bellmanFord(const std::string &source, const std::string &target) {
+    for (Vertex *v: vertexSet) {
+        v->setCost(UINT32_MAX);
+        v->setPath(nullptr);
+    }
+    findVertex(source)->setCost(0);
+
+    for (int i = 1; i <= vertexSet.size(); i++) { //V times
+        for (Vertex *v: vertexSet) { //Relax every Edge
+            for (Edge *e: v->getIncoming()) {
+                if (e->getCapacity() > 0) {
+                    int tempCost = e->getOrig()->getCost() + e->getCost();
+                    if (tempCost < v->getCost()) {
+                        if (i == vertexSet.size()) { //Edge being relaxed on Nth iteration - Negative cycle!
+                            Vertex *currentVertex = v;
+                            std::list<Edge *> negativeCycle;
+                            do {
+                                negativeCycle.push_back(currentVertex->getPath());
+                                currentVertex = currentVertex->getPath()->getOrig();
+                            } while (currentVertex != v);
+                            return negativeCycle;
+                        }
+                        v->setCost(tempCost);
+                        v->setPath(e);
+                    }
+                }
+            }
+        }
+    }
+    return {};
+}
+
+std::pair<unsigned int, unsigned int>
+Graph::minCostMaxFlow(const std::string &source, const std::string &target, Graph &residualGraph) {
+    std::pair<unsigned int, unsigned int> result;
+    result.first = edmondsKarp({source}, target, residualGraph);
+
+    Graph minCostResidual;
+    makeMinCostResidual(minCostResidual);
+
+    std::list<Edge *> negativeCycle = minCostResidual.bellmanFord(source, target);
+    while (!negativeCycle.empty()) {
+        unsigned int bottleneckCapacity = minCostResidual.findListBottleneck(negativeCycle);
+        minCostResidual.augmentMinCostPath(negativeCycle, bottleneckCapacity);
+        negativeCycle = bellmanFord(source, target);
```

Comment:
I'm not following, we update the residual then check the original?

Reply:
That's a typo, should be minCostResidual.bellmanFord